### PR TITLE
store/tikv: update scanner

### PR DESCRIPTION
**Pull request by disksing**
@shenli @qiuyesuifeng @morefreeze

### CODE REVIEW

File: store/tikv/scan_mock_test.go [hunk 1]
```diff
@@ -53,28 +51,3 @@ func (s *testScanMockSuite) TestScanMultipleRegions(c *C) {
 	c.Assert(scanner.Next(), NotNil)
 	c.Assert(scanner.Valid(), IsFalse)
 }
-
-func (s *testScanMockSuite) TestStaleRegionEpoch(c *C) {
```
Comment (morefreeze) [inline store/tikv/scan_mock_test.go]: Why remove this?

Comment (disksing) [inline store/tikv/scan_mock_test.go]: @morefreeze We don't need it any more after adding retry for region error.

File: store/tikv/scan.go [hunk 2]
```diff
@@ -87,117 +79,126 @@ func (s *Scanner) Next() error {
...
+			if s.eof {
+				s.Close()
+				return kv.ErrNotExist
+			}
+			err := s.getData()
+			if err != nil {
+				s.Close()
+				return errors.Trace(err)
+			}
+			if s.idx >= len(s.cache) {
```
Comment (qiuyesuifeng) [inline store/tikv/scan.go]: Can u show a detail why check this condition?

File: store/tikv/scan.go [hunk 3]
```diff
@@ -87,117 +79,126 @@ func (s *Scanner) Next() error {
...
+
+		// Update nextStartKey and skipFirst.
+		if s.mayHasMoreRegionData() {
+			lastPair := kvPairs[len(kvPairs)-1]
+			s.nextStartKey, s.skipFirst = lastPair.GetKey(), true
+		} else if len(region.EndKey()) == 0 {
+			// No data in current Region, and no Region after this Region.
+			s.eof = true
+		} else {
+			s.nextStartKey, s.skipFirst = region.EndKey(), false
```
Comment (qiuyesuifeng) [inline store/tikv/scan.go]: Can we use helper function like `nextKey` instead of `skipFirst` logic?

File: store/tikv/scan.go (hunk 2 above)
Comment (disksing) [inline store/tikv/scan.go]: @qiuyesuifeng This could happen if there is no more data in current Region.

File: store/tikv/scan.go (hunk 3 above)
Comment (disksing) [inline store/tikv/scan.go]: @qiuyesuifeng Agree. I'm going to have a try.

**Comment by disksing** [comment]
PTAL @qiuyesuifeng

**Comment by morefreeze** [comment]
Did this scanner have test that key is deleting?

**Comment by qiuyesuifeng** [comment]
LGTM

File: store/tikv/scan.go [hunk 4]
```diff
@@ -87,128 +77,124 @@ func (s *Scanner) Next() error {
...
+		resp, err := s.snapshot.store.SendKVReq(req, region)
 		if err != nil {
-			return nil, errors.Trace(err)
+			return errors.Trace(err)
 		}
-		ret[i] = &resultRow{
-			key:  lockInfo.GetKey(),
-			lock: lockInfo,
+		if regionErr := resp.GetRegionError(); regionErr != nil {
+			log.Warnf("scanner getData failed: %v", regionErr.String())
```
Comment (morefreeze) [inline store/tikv/scan.go]: Why not using `log.Warnf("xxx %s", regionErr)`

**Comment by morefreeze** [comment]
LGTM

**Pull request merged**